### PR TITLE
Add shutdown to EventListener

### PR DESCRIPTION
**Pull request by dominikzalewski**
## Description

@PreDestroy and @PostCreate annotations are processed on EventListener methods.

## Release notes

( X ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java [hunk 1]
```diff
@@ -83,6 +86,7 @@ public void addEventListener(EventListener eventListener)
         requireNonNull(eventListener, "EventListener is null");
 
         providedEventListeners.add(eventListener);
+        lifeCycleManager.addInstance(eventListener);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:89]: i think it's not unreasonable for `io.trino.spi.eventlistener.EventListener` to be informed about system shutdown

if we want that, this should be an explicit method on that interface though, not some implicit annotation-reading which is hard to discover by interface implementors, and may not even work across classloaders.

Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:89]: Let's follow `io.trino.spi.connector.Connector#shutdown` practice

File: core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java [hunk 2]
```diff
@@ -62,11 +63,13 @@ public class EventListenerManager
     private final TimeStat queryCreatedTime = new TimeStat(MILLISECONDS);
     private final TimeStat queryCompletedTime = new TimeStat(MILLISECONDS);
     private final TimeStat splitCompletedTime = new TimeStat(MILLISECONDS);
+    protected final LifeCycleManager lifeCycleManager;
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:66]: why is it protected?

File: core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java (hunk 1 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:89]: Yes. It'd be better to dispatch the shutdown event from the `EventListenerManager`. It should have @PreDestroy method that will call all event listeners shutdown method

**Review by wendigo — CHANGES_REQUESTED** [review]

File: core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java (hunk 1 above)
Comment (electrum) [reply core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:89]: This wouldn't work anyway, as the annotations in the plugin are in a different class loader than the engine.

**Review by electrum — CHANGES_REQUESTED** [review]
We should add an explicit `shutdown` method, like `Connector`

File: core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java (hunk 2 above)
Comment (dominikzalewski) [reply core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:66]: I needed that in TestingLifecycleManager that inherits. But this is going to change now anyway.

**Comment by dominikzalewski** [comment]
Changed as requested:
- EventListener.shutdown() also calls `close` if it is `AutoCloseable` -> let me know whether that's OK with you. If not, I'll remove
- Added unit tests to verify that the call is forwarded from manager to listener

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java [hunk 3]
```diff
@@ -34,4 +34,12 @@ default boolean requiresAnonymizedPlan()
     {
         return false;
     }
+
+    default void shutdown()
+            throws Exception
```
Comment (wendigo) [inline core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java:39]: I'd just make it:
```
default void shutdown() {}
```

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java [hunk 4]
```diff
@@ -34,4 +34,12 @@ default boolean requiresAnonymizedPlan()
...
+    {
+        if (this instanceof AutoCloseable autoCloseable) {
```
Comment (wendigo) [inline core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java:41]: AutoCloseable is not needed. Event Listener isn't used in try-with-resources block anyway

**Review by wendigo — APPROVED** [review]

**Comment by wendigo** [comment]
And this failure is related:

```
java.lang.AssertionError: io.trino.plugin.base.classloader.ClassLoaderSafeEventListener does not override [public default void io.trino.spi.eventlistener.EventListener.shutdown() throws java.lang.Exception]
```

File: core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java [hunk 5]
```diff
@@ -218,4 +219,17 @@ public TimeStat getSplitCompletedTime()
     {
         return splitCompletedTime;
     }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        for (EventListener listener : configuredEventListeners.get()) {
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:226]: If we are shutting down event listener manager than I think we should actually make it deconfigure existing listeners.

This line should be:

```
for (EventListener listener : configuredEventListeners.getAndSet(List.of())) {
```

which returns previously configured event listeners and swaps them with empty list.

@electrum @findepi @kokosing wdyt?

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java [hunk 6]
```diff
@@ -34,4 +34,8 @@ default boolean requiresAnonymizedPlan()
     {
         return false;
     }
+
+    default void shutdown()
```
Comment (wendigo) [inline core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java:38]: nit:

```
/**
 * Shutdown the event listener by releasing any held resources such as
 * threads, sockets, etc. After this method is called,
 * no methods will be called on the event listener.
 */
default void shutdown() {}
```

**Comment by wendigo** [comment]
LGTM @dominikzalewski. Thanks!

**Pull request merged**